README.md

```diff
@@ -278,167 +310,146 @@ representation:
   try {
-    const x = $$try1.value;
-    {
-      const $$try2 = { value: expr2, hasError: false, error: undefined };
-      try {
-        const y = $$try1.value;
-        ...
-      }
-      catch ($$error) {
-        $$try2.hasError = true;
-        $$try2.error = $$error;
-        throw $$error;
-      }
-      finally {
-        try {
-          if ($$try2.value !== null && $$try2.value !== undefined) {
-            $$try2.value[Symbol.dispose]();
-          }
-        }
-        catch ($$error) {
-          if ($$try2.hasError) {
-            throw new AggregateError([$$try2.error, $$error]);
-          }
-          throw $$error;
-        }
-      }
+    const x = expr1;
+    if (x !== null && x !== undefined) {
+      const $$dispose = x[Symbol.dispose];
+      if (typeof $$dispose !== "function") throw new TypeError();
+      $$try.stack.push({ value: x, dispose: $$dispose });
+    }
+
+    const y = expr2;
+    if (y !== null && y !== undefined) {
+      const $$dispose = y[Symbol.dispose];
+      if (typeof $$dispose !== "function") throw new TypeError();
+      $$try.stack.push({ value: y, dispose: $$dispose });
     }
+
+    ...
+
   }
   catch ($$error) {
-    $$try1.hasError = true;
-    $$try1.error = $$error;
-    throw $$error;
+    $$try.errors.push($$error);
   }
   finally {
-    try {
-      if ($$try1.value !== null && $$try1.value !== undefined) {
-        $$try1.value[Symbol.dispose]();
+    while ($$try.stack.length) {
+      const { value: $$expr, dispose: $$dispose } = $$try.stack.pop();
+      try {
+        $$dispose.call($$expr);
       }
-    }
-    catch ($$error) {
-      if ($$try1.hasError) {
-        throw new AggregateError([$$try1.error, $$error]);
+      catch ($$error) {
+        $$try.errors.push($$error);
       }
-      throw $$error;
+    }
+    if ($$try.errors.length > 1) {
+      throw new AggregateError($$try.errors);
+    }
+    if ($$try.errors.length === 1) {
+      throw $$try.errors[0];
     }
   }
 }
 ```
 
 Since we must always ensure that we properly release resources, we must ensure that any abrupt 
 completion that might occur during binding initialization results in evaluation of the cleanup 
-step. This also means that when there are multiple declarations in the list we must create a 
-new `try/finally`-like protected region for each declaration. As a result, we must release 
-resources in reverse order.
+step. When there are multiple declarations in the list, we track each resource in the order they 
+are declared. As a result, we must release these resources in reverse order.
 
-## `try` with binding patterns
+## `try`-`using` with binding patterns
 
-The `try` statement always creates implicit local bindings for the _Initializer_ of the 
-_LexicalBinding_. For binding patterns this means that we store the value 
-of `expr` in the example below, rather than `y`:
+The `try`-`using` statement tracks disposable resources for each named binding, in the order the bindings
+are evaluated:
```

Comment:
Hi.

If there is some function S which returns an object with resources a, b and c, then if I bind two of them, then only those two of them are going to be disposed:
```js
function S() {
  return {
    a: createResource('a'),
    b: createResource('b'),
    c: createResource('c'),
  }
}

try using (const { a, b } = S()) {
  ...
}
```
I think that should be error-prone.
Disposing only bound values moves control over what will be disposed from the producer of a disposable object to the consumer of that object.

The only error-free implementation is to produce an object, which generates resources on demand in property getters. But imho it seems to be a kind of hack.

Consider previous behavior, when the entire expression is disposed in `try using`:
```js
function S() {
  return {
    a: createResource('a'),
    b: createResource('b'),
    c: createResource('c'),
    [Symbol.dispose]() {
      this.a[Symbol.dispose]()
      this.b[Symbol.dispose]()
      this.c[Symbol.dispose]()
    },
  }
}

try using (const { a, b } = S()) {
  ...
}
```
In this case all the three resources will be destroyed and the code seems to be more explicit.
Also, returned object is allowed to have e.g. non-disposable field 'd', and we are not able to bind it if we're disposing bound variables instead of the expression.

The disposing-expression behavior is also implemented in C++.
If some object is returned from a function, then the destructor of that object, not destructors of its parts, will be called when leaving the scope.
So I don't expect that disposing expression, but not bound variables, could lead to much misunderstanding.